### PR TITLE
Optimizations for update() method inside mesh._draw

### PR DESCRIPTION
**Pull request by asobolev**

### CODE REVIEW

File: ratcave/mixins.py [hunk 1]
```diff
@@ -71,16 +73,29 @@ def __init__(self, position=(0., 0., 0.), rotation=(0., 0., 0.), scale=1., *args
             scale (float): uniform scale factor. 1 = no scaling.
         """
         super(Physical, self).__init__(*args, **kwargs)
-        self.x, self.y, self.z = position
-        self.rot_x, self.rot_y, self.rot_z = rotation
+        self.__dict__['x'], self.__dict__['y'], self.__dict__['z'] = position
```
Comment (neuroneuro15) [inline ratcave/mixins.py]: Why the change to using self.__dict__?

**Comment by neuroneuro15** [comment]
Hi Andrey,

From what I understand, this commit has three main changes:
  1) Mesh.update() now only gets called the first time it is drawn, not every time (Performance optimization).
  2) Physical objects call self./_/_dict/_/_ to get their attributes (Performance Optimization?)
  3) min_xyz and max_xyz now have the corner boundaries of the object, for 3d collision detection. (New Feature)

Most of it looks good, but I'm concerned that moving objects won't appear correct if scripts forget to call Mesh.update() explicitly whenever the mesh, or its parents, are moved.  Do I understand this correctly, or am I missing something? Maybe some middle ground would be better (i.e. if Mesh.update() gets called whenever some attribute gets changed).  Could you take a look at this and comment before I merge the pull request?

Best wishes,

Nick

**Comment by asobolev** [comment]
Hi Nick,

thanks a lot for looking at the pull request.

> 1) Mesh.update() now only gets called the first time it is drawn, not
> every time (Performance optimization).
>
Right, and also every time any of the position / shape attributes are
changes (__observables = ('x', 'y', 'z', 'rot_x', 'rot_y', 'rot_z',
'scale')). Such that the object is updated ONLY if it is changed. Would
that be a correct strategy?

The implementation of this feature is the shortest possible (overwriting a
__setattr__ method etc.). However it results in an ugly code like in
__init__ like self.__dict__[<smth>] = bla which is not the coolest
practice. A good implementation will be longer, so I keep it as it is right
now. But if you say you want it the "right" way I can do it too.


> 2) Physical objects call self./*/_dict/*/_ to get their attributes
> (Performance Optimization?)
>
This is an implementation feature of the above. It avoids calling the
__setattr__ method to assign initial values for "observed" variables.

> 3) min_xyz and max_xyz now have the corner boundaries of the object, for
> 3d collision detection. (New Feature)
>
Exactly. That's the easiest solution (rectangular) for now. We can
elaborate that in the future, of course.


Thanks again for looking the the request. Let me know if something is not
clear!

Andrey

File: ratcave/mixins.py (hunk 1 above)
Comment (asobolev) [inline ratcave/mixins.py]: This is an implementation feature of the "observed" variables (__observables = ('x', 'y', 'z', 'rot_x', 'rot_y', 'rot_z', 'scale')). It avoids calling the __setattr__ method to assign initial values for them.

**Comment by asobolev** [comment]
Hi Nick, I updated the master branch so there is no conflict anymore. Please check it up and merge if it's fine with you. Thanks!

**Comment by neuroneuro15** [comment]
Looks great, thanks for doing this again!  Merging!
